lib/mongo/server/connection_pool/queue.rb

```diff
@@ -165,12 +164,48 @@ def wait_timeout
           @wait_timeout ||= options[:wait_queue_timeout] || WAIT_TIMEOUT
         end
 
+        # The maximum seconds a socket can remain idle since it has been checked in to the pool.
+        #
+        # @example Get the max idle time.
+        #   queue.max_idle_time
+        #
+        # @return [ Float ] The max socket idle time in seconds.
+        #
+        # @since 2.5.0
+        def max_idle_time
+          @max_idle_time ||= options[:max_idle_time]
+        end
+
+        # Close sockets that have been open for longer than the max idle time, if the
+        #   option is set.
+        #
+        # @example Close the stale sockets
+        #   queue.close_stale_sockets!
+        #
+        # @since 2.5.0
+        def close_stale_sockets!
+          if max_idle_time
+            mutex.synchronize do
+              queue.reverse_each do |connection|
+                if last_checkin = connection.last_checkin
+                  if (Time.now - last_checkin) > max_idle_time
+                    connection.disconnect!
+                    if queue.index(connection) < min_size
+                      connection.connect!
+                    end
```

Comment:
Does `connection.connect!` block? In Python we have a rule: no blocking I/O while holding a lock. Does Ruby follow a similar mantra?